src/installer.rs

```diff
@@ -28,123 +28,165 @@ pub struct ServerInstallation {
 pub struct MinecraftVersion {
     pub version: String,
-    pub stable: bool
+    pub stable: bool,
 }
 
-impl Display for MinecraftVersion {
+impl std::fmt::Display for MinecraftVersion {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         write!(f, "{}", self.version)
     }
 }
 
-
 #[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
 pub struct LoaderVersion {
-    pub separator: String,
+    pub separator: char,
     pub build: u32,
     pub maven: String,
-    pub version: String,
+    pub version: Version,
 }
 
-impl Display for LoaderVersion {
+impl std::fmt::Display for LoaderVersion {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         write!(f, "{}", self.version)
     }
 }
 
 pub async fn fetch_minecraft_versions() -> Result<Vec<MinecraftVersion>> {
-    Ok(reqwest::get("https://meta.quiltmc.org/v3/versions/game").await?.json().await?)
+    Ok(get("https://meta.quiltmc.org/v3/versions/game")
+        .await?
+        .json()
+        .await?)
 }
 
 pub async fn fetch_loader_versions() -> Result<Vec<LoaderVersion>> {
-    Ok(reqwest::get("https://meta.quiltmc.org/v3/versions/loader").await?.json().await?)
+    Ok(get("https://meta.quiltmc.org/v3/versions/loader")
+        .await?
+        .json()
+        .await?)
 }
 
-#[derive(Serialize, Deserialize)]
-struct LaunchProfiles {
-    profiles: HashMap<String, serde_json::Value>,
-    settings: serde_json::Value,
-    version: u32
+/// `Deserialize` is not implemented for a reason
+///
+/// DO NOT deserialise `launcher_profiles.json` into this incomplete struct and write it back as it will cause **data loss**
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+struct Profile {
+    name: String,
+    #[serde(rename = "type")]
+    profile_type: String,
+    created: DateTime<Utc>,
+    last_version_id: String,
+    icon: String,
 }
 
 pub async fn install_client(args: ClientInstallation) -> Result<()> {
     println!("Installing client: {:#?}", args);
 
     // Verify install location
     if !args.install_location.exists() {
-        return Err(anyhow!("Target directory doesn't exist: {:?}", args.install_location));
+        return Err(anyhow!(
+            "Target directory doesn't exist: {}",
+            args.install_location.display(),
+        ));
     }
 
     // Resolve profile directory
-    let profile_name = format!("quilt-loader-{}-{}", args.loader_version.version, args.minecraft_version.version);
-    let mut profile_dir = args.install_location.clone();
-    profile_dir.push("versions");
-    profile_dir.push(&profile_name);
+    let profile_name = format!("quilt-loader-{}-{}", args.loader_version, args.minecraft_version);
+    let profile_dir = args.install_location.join("versions").join(&profile_name);
 
     // Delete existing profile
     if profile_dir.exists() {
-        std::fs::remove_dir_all(&profile_dir)?;
+        fs::remove_dir_all(&profile_dir)?;
```

Comment:
This is actually the `.minecraft/versions` folder for the specific version, and not the launcher profile (perhaps it could do with a rename?) I'm pretty sure it doesn't matter if it is deleted, what do you think?

Reply:
Yes, I agree in principle.
However, we still continue to *create* the directory, even if we aren't in a valid mc installation.
In either case, verifying the directory let's us abort gracefully with a nice error message, rather than doing stuff that *shouldn't* be a problem.

Reply:
Actually yeah I forgot. There was another another bug cos the installer just assumed `launcher_profiles.json` exists. A check will fix that too

Reply:
4465b6d4c9d264a6efeaf2d4e23ba5dd10a2e395